core/src/main/java/feign/FeignException.java

```diff
@@ -40,58 +41,70 @@
   private static final long serialVersionUID = 0;
   private int status;
   private byte[] responseBody;
+  private Map<String, Collection<String>> responseHeaders;
   private Request request;
 
   protected FeignException(int status, String message, Throwable cause) {
     super(message, cause);
     this.status = status;
+    this.responseHeaders = Collections.emptyMap();
     this.request = null;
   }
 
-  protected FeignException(int status, String message, Throwable cause, byte[] responseBody) {
+  protected FeignException(int status, String message, Throwable cause, byte[] responseBody,
+      Map<String, Collection<String>> responseHeaders) {
     super(message, cause);
     this.status = status;
     this.responseBody = responseBody;
+    this.responseHeaders = responseHeaders;
```

Comment:
Despite the fact that Strings are immutable, the Map is not.  Can you change this to create a copy of the map?  Some examples are `Map.of(responseHeaders)` or `new LinkedHashMap<>(responseHeaders)`

Reply:
Will do. I'm just wondering if wrapping the map itself is enough? The values of this map are of type ``Collection<String>``, so although wrapping the map would make it immutable, the values would still be mutable. To make this "truly" immutable, we would need to wrap the values as well. Or do you think that's an overkill?

Reply:
Strings are immutable by nature, so to ensure immutability, you must lock down the map.

Reply:
True that but having an immutable map still gives you the option of `responseHeaders().get("Something").clear()` unless `Collection<String>` isn't also wrapped in an `Collections.unmodifiableCollection()` at some point before.

Reply:
Response has the method `caseInsensitiveCopyOf`. I think it should be moved to `Utll` and reused here.